anti-diff/src/Data/Map/Diff/Strict/Internal.hs

```diff
@@ -0,0 +1,353 @@
+    m1
+    m2
+
+fromList :: Ord k => [(k, DiffHistory v)] -> Diff k v
+fromList = Diff . Map.fromList
+
+fromListInserts :: Ord k => [(k, v)] -> Diff k v
+fromListInserts = fromList . fmap (second singletonInsert)
+
+fromListDeletes :: Ord k => [(k, v)] -> Diff k v
+fromListDeletes = fromList . fmap (second singletonDelete)
+
+singleton :: DiffEntry v -> DiffHistory v
+singleton = DiffHistory . Seq.singleton
+
+singletonInsert :: v -> DiffHistory v
+singletonInsert = singleton . Insert
+
+singletonDelete :: v -> DiffHistory v
+singletonDelete = singleton . Delete
+
+fromSeq :: DiffEntry v -> Seq (DiffEntry v) -> DiffHistory v
+fromSeq x xs = DiffHistory $ x Seq.:<| xs
+
+{------------------------------------------------------------------------------
+  Class instances for @'Diff'@
+------------------------------------------------------------------------------}
+
+-- Note: The use of @'isNonEmptyHistory'@ prevents the merged @'Diff'@s from
+-- getting bloated with empty @'DiffHistory'@s.
+instance (Ord k, Eq v) => Semigroup (Diff k v) where
+  Diff m1 <> Diff m2 = Diff $
+    Merge.merge
+      Merge.preserveMissing
+      Merge.preserveMissing
+      (Merge.zipWithMaybeMatched
+        (\_k h1 h2 -> isNonEmptyHistory $ h1 <> h2)
+      )
+      m1
+      m2
+
+instance (Ord k, Eq v) => Monoid (Diff k v) where
+  mempty = Diff mempty
+
+instance (Ord k, Eq v) => Group (Diff k v) where
+  invert (Diff m) = Diff $ fmap invert m
+
+{------------------------------------------------------------------------------
+  Class instances for @'DiffHistory'@
+------------------------------------------------------------------------------}
+
+-- | @h1 <> h2@ sums @h1@ and @h2@ by cancelling out as many consecutive diff
+-- entries as possible.
+--
+-- Diff entries that are each other's inverse can cancel out. In this case,
+-- both diff entries are removed the diff history.
+--
+-- Examples:
+-- > [Insert 1, AntiDelete 2] <> [Delete 2, AntiInsert 1] = []
+-- > [Insert 1, Delete 2]     <> [AntiDelete 2, Delete 3] = [Insert 1, Delete 3]
+-- > [Insert 1, AntiInsert 1] <> []                       = [Ins 1, AntiInsert 1]
+--
+-- Note: This implementation does not make any assumptions about the inputs it is
+-- given, and the results it produces. Because of this, the resulting diff
+-- history can be nonsenical. To illustrate, consider example 2 given above:
+-- how could we expect to delete a value @3@ if we inserted a value @1@? As
+-- such, it is the using code's responsibility to ensure that the inputs given
+-- to the sum lead to sensible results.
+--
+-- Note: We do not cancel out consecutive elements in @h1@ and @h2@
+-- individually. It is only at the border between @h1@ and @h2@ that we cancel
+-- out elements (see the third example given above).
+instance Eq v => Semigroup (DiffHistory v) where
+  DiffHistory s1 <> DiffHistory s2 = DiffHistory $ s1 `mappend'` s2
+    where
+      -- At the ``touching'' ends of the sequences, take off diff entries that
```

Comment:
It is an optimisation because the aim was that combining inserts/deletes with anti-inserts/anti-deletes should make the diff history smaller, like it's subtracting elements. Just from the top of my head, it is also essential to its correctness because it must satisfy the `Group` laws: `invert x <> x == mempty` and `x <> invert x == mempty`